datalad_next/runners/data_processors.py

```diff
@@ -0,0 +1,316 @@
+    """ Split character-strings or byte-strings into lines
+
+    Parameters
+    ----------
+    data_chunks: list[str | bytes]
+        a list of strings or bytes
+    final : bool
+        the data chunks are the final data chunks of the source. A line is
+        terminated by end of data.
+
+    Returns
+    -------
+    list[str | bytes]
+        if the input data chunks contained bytes the result will be a list of
+        byte-strings that end with byte-size line-delimiters. If the input data
+        chunks contained strings, the result will be a list strings that end with
+        string delimiters (see Python-documentation for a definition of string
+        line delimiters).
+    """
+    # We use `data_chunks[0][0:0]` to get an empty value the proper type, i.e.
+    # either the string `''` or the byte-string `b''`.
+    empty = data_chunks[0][0:0]
+    text = empty.join(data_chunks)
+    # Use the builtin line split-wisdom of Python
+    parts_with_ends = text.splitlines(keepends=True)
+    parts_without_ends = text.splitlines(keepends=False)
+    if parts_with_ends[-1] == parts_without_ends[-1] and not final:
+        return parts_with_ends[:-1], [parts_with_ends[-1]]
+    return parts_with_ends, []
+
+
+def _pattern_border_processor(pattern: StrOrBytes,
+                              data_chunks: StrOrBytesList,
+                              final: bool = False,
+                              ) -> tuple[StrOrBytesList, StrOrBytesList]:
+
+    def ends_with_pattern_prefix(data: StrOrBytes, pattern: StrOrBytes) -> bool:
+        """ Check whether the chunk ends with a prefix of the pattern """
+        for index in range(len(pattern) - 1, 0, -1):
+            if  data[-index:] == pattern[:index]:
+                return True
+        return False
+
+    # Copy the list, because we might modify it and the caller might not expect that.
+    data_chunks = data_chunks[:]
+
+    # Join data chunks until they are sufficiently long to contain the pattern,
+    # i.e. have a least the size: `len(pattern)`. Continue joining, if the chunk
+    # ends with a prefix of the pattern.
+    current_index = 0
+    while current_index < len(data_chunks) - 1:
+        current_chunk = data_chunks[current_index]
+        while (len(data_chunks[current_index:]) > 1
+               and (len(current_chunk) < len(pattern)
+                    or ends_with_pattern_prefix(current_chunk, pattern))):
+            data_chunks[current_index] += data_chunks[current_index + 1]
+            del data_chunks[current_index + 1]
+            current_chunk = data_chunks[current_index]
+        current_index += 1
+
+    # At this point we have joined whatever we can join. We still have to check
+    # whether the last chunk ends with a pattern-prefix.
+    if not final:
+        if ends_with_pattern_prefix(data_chunks[-1], pattern):
+            return data_chunks[:-1], data_chunks[-1:]
+    return data_chunks, []
+
+
+class SplitLinesProcessor:
```

Comment:
Undocumented

Reply:
I removed the class `SplitLinesProcessor`. The `LineSplitter` that was used by `SplitLinesProcessor` has shortcomings that make it not feasible for the pipelines, i.e. it does not support `bytes`  and, as a minor issue, it cannot be properly re-used.